core/src/main/java/org/apache/iceberg/encryption/StandardEncryptionManager.java

```diff
@@ -102,106 +134,69 @@ private SecureRandom workerRNG() {
   }
 
   public ByteBuffer wrapKey(ByteBuffer secretKey) {
-    if (kmsClient == null) {
+    if (keyData == null) {
       throw new IllegalStateException(
           "Cannot wrap key after called after serialization (missing KMS client)");
     }
 
-    return kmsClient.wrapKey(secretKey, tableKeyId);
+    return keyData.kmsClient.wrapKey(secretKey, tableKeyId);
   }
 
   public ByteBuffer unwrapKey(ByteBuffer wrappedSecretKey) {
-    if (kmsClient == null) {
+    if (keyData == null) {
       throw new IllegalStateException("Cannot unwrap key after serialization (missing KMS client)");
     }
 
-    return kmsClient.unwrapKey(wrappedSecretKey, tableKeyId);
+    return keyData.kmsClient.unwrapKey(wrappedSecretKey, tableKeyId);
   }
 
-  void addKekCache(Map<String, KeyEncryptionKey> wrappedKekCache) {
-    for (Map.Entry<String, KeyEncryptionKey> entry : wrappedKekCache.entrySet()) {
-      KeyEncryptionKey wrappedKek = entry.getValue();
-      KeyEncryptionKey cachedKek = kekCache.get(entry.getKey());
-
-      if (cachedKek != null) {
-        Preconditions.checkState(
-            cachedKek.wrappedKey().equals(wrappedKek.wrappedKey()),
-            "Cached kek wrap differs from newly added for %s",
-            entry.getKey());
-      } else {
-        ByteBuffer encryptedKEK =
-            ByteBuffer.wrap(Base64.getDecoder().decode(wrappedKek.wrappedKey()));
-        // Unwrap the key in KMS
-        byte[] kekBytes = unwrapKey(encryptedKEK).array();
-
-        kekCache.put(
-            entry.getKey(),
-            new KeyEncryptionKey(
-                wrappedKek.id(), kekBytes, wrappedKek.wrappedKey(), wrappedKek.timestamp()));
-      }
+  public String currentSnapshotKeyId() {
+    if (keyData == null) {
+      throw new IllegalStateException("Cannot return the current snapshot key after serialization");
     }
-  }
-
-  KeyEncryptionKey currentKEK() {
-    if (kekCache.isEmpty()) {
-      KeyEncryptionKey keyEncryptionKey = generateNewKEK();
-      kekCache.put(keyEncryptionKey.id(), keyEncryptionKey);
 
-      return keyEncryptionKey;
+    if (keyData.currentEncryptionKey == null
+        || keyData.currentEncryptionKey.timestamp()
+            < System.currentTimeMillis() - writerKekTimeout) {
+      createNewEncryptionKey();
     }
 
-    KeyEncryptionKey lastKek = lastKek();
-    long timeNow = System.currentTimeMillis();
-    if (timeNow - lastKek.timestamp() > writerKekTimeout) {
-      KeyEncryptionKey keyEncryptionKey = generateNewKEK();
-      kekCache.put(keyEncryptionKey.id(), keyEncryptionKey);
+    return keyData.currentEncryptionKey.id();
+  }
 
-      return keyEncryptionKey;
+  public ByteBuffer unwrapKey(String keyId) {
+    if (keyData == null) {
+      throw new IllegalStateException("Cannot unwrap key after serialization (missing KMS client)");
     }
 
-    return lastKek;
+    return unwrapKey(keyData.encryptionKeys.get(keyId).wrappedKey());
```

Comment:
Keeping track of keys and caching the `unwrapKey` operation are two separate things. Since the two were mixed together, I've removed the caching for now.

The next step is to create a caffeine `Cache` that handles unwrapping here.